libdnf/base/base.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/base/base.hpp"
+
+#include "libdnf/conf/config_parser.hpp"
+#include "libdnf/conf/const.hpp"
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <filesystem>
+#include <vector>
+
+namespace libdnf {
+
+static void load_config_from_parser(
+    Config & conf, const ConfigParser & parser, const std::string & section, Logger & log) {
+    const auto & cfg_parser_data = parser.get_data();
+    auto cfg_parser_data_iter = cfg_parser_data.find(section);
+    if (cfg_parser_data_iter != cfg_parser_data.end()) {
+        auto opt_binds = conf.opt_binds();
+        const auto & cfg_parser_main_sect = cfg_parser_data_iter->second;
+        for (const auto & opt : cfg_parser_main_sect) {
+            auto opt_binds_iter = opt_binds.find(opt.first);
+            if (opt_binds_iter != opt_binds.end()) {
+                try {
+                    opt_binds_iter->second.new_string(libdnf::Option::Priority::MAINCONFIG, opt.second);
+                } catch (const Option::Exception & ex) {
+                    auto msg = fmt::format(
+                        R"**(Config error in section "{}" key "{}": {}: {})**",
+                        section,
+                        opt.first,
+                        ex.get_description(),
+                        ex.what());
+                    log.warning(msg);
+                }
+            }
+        }
+    }
+}
+
+static void load_config_from_file_path(
+    Config & conf, const std::string & file, const std::string & section, Logger & log) {
+    log.debug(fmt::format(R"**(Start of loading section "{}" from file "{}")**", section, file));
+    ConfigParser parser;
+    parser.read(file);
+    load_config_from_parser(conf, parser, section, log);
+    log.debug(fmt::format(R"**(Loading of section "{}" from file "{}" done)**", section, file));
+}
+
+void Base::load_config_from_file(const std::string & path) {
+    load_config_from_file_path(config, path, "main", log_router);
+}
```

Comment:
In past it was possible to add a definition of repositories from conf file. I suggest that we have to support it for compatibility reasons. Personally I don't like it, but the compatibility is the important element.

Reply:
Yes. Repositories defined in main configuration file will be supported for backward compatibility. But PR is only about loading main configuration. I have code for loading repositories configuration too. It will be on github in next days. But the code for loading rpm repositories configuration is "independent" of code for loading main configuration.
OK, there is one side effect. The main configuration file is opened/readed twice. During reading main configuration and when repositories configuration is read. On the other hand it will be very simple to activate/deactivate loading repositories from main configuration file.